api/http-api.md

```diff
@@ -0,0 +1,177 @@
+
+For the initial version, we will only be doing workflow service. Operator service and cloud APIs may come soon after
+though.
+
+### Rules
+
+* Naming
+  * Multi-word URL paths use kebab-case
+  * Multi-word query parameters use camelCase (only first letter of abbreviations capitalized if not first word)
+  * Do not assume any names reserved. So we can't have `/workflows/{workflow_id}` and `/workflows/general-call` in case
+    there's a workflow named `general-call`.
+  * Use plural nouns where an identifier may be provided to differentiate
+    * But singular nouns are not _required_ in other places, can still be plural
+* Response codes
+  * 201 success if a creation
+  * 202 if action eventually consistent
+  * 200 other all other successes
+  * Do not use 204 for no response, even if there's no response because we may add response later, just 200 everything
+* Methods:
+  * `GET` if read-only
+    * Noun as last URL term
+    * Use query parameters for commonly set items to avoid relying on request bodies (but still rely on bodies as
+      needed). This may require some gRPC gateway customization.
+  * `PUT` if strongly consistent add/update on fixed-URL w/ idempotency
+    * Noun as last URL term
+  * `DELETE` if item is actually deleted (i.e. unreachable from `GET`)
+    * Noun as last URL term
+  * `POST` for everything else
+    * Noun as last URL term if for a create, verb if for an action
+  * `PATCH` is not needed because we don't adhere to common CRUD
+    * TODO(cretz): Needs discussion
+
+### RPCs
+
+Below is list of gRPC calls to HTTP calls so far and some notes/questions beneath each. Some notes/questions beneath one
+actually apply to many but we are not repeating ourselves too much here.
+
+* Namespaces:
+  * `RegisterNamespace` - `POST /api/v1/namespaces`
+  * `DescribeNamespace` - `GET /api/v1/namespaces/{namespace}`
+  * `ListNamespaces` - `GET /api/v1/namespaces`
+  * `UpdateNamespace` - `PUT /api/v1/namespaces/{namespace}`
+    * TODO(cretz): Confirm strongly consistent and idempotent or not
+  * `DeprecateNamespace` - `DELETE /api/v1/namespaces/{namespace}`
+    * Is this eventually consistent? Will determine response code
+    * Does this prevent returning from `GET`? Will determine HTTP method
+* Workflows:
+  * `StartWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}`
+    * May accept header as `identity` and `request_id` on these calls
+  * `GetWorkflowExecutionHistory` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history`
+    * `?nextPageToken` can be a query param even though it's big
+  * `GetWorkflowExecutionHistoryReverse` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history-reverse`
+  * `RequestCancelWorkflowExecutionRequest` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/cancel`
+    * Intentionally not taking the run ID in the URL, but should it be `?runId=` query param?
```

Comment:
Yes, but it is optional. I think of it more like a filter than a resource. Having said that, yes, we could have a separate `runs` endpoint, but with these APIs I am trying not to have two URL endpoints mean the same thing. I could even have a `/latest` for absent run. Also not trying to have `/runs` for all of these.

But it's subjective and it isn't that ugly if others agree. Always easy to add later too.

Reply:
I don't really like `/latest` because I think it violates `Do not assume any names [are] reserved`, but that's fair. If it's something that we expect people to use very often, I'd even be fine with leaving it out entirely and deciding later if we want `?runId=...` or a separate route.

Reply:
> I don't really like /latest because I think it violates Do not assume any names [are] reserved

It would be `/api/v1/namespaces/{namespace}/workflows/{workflow_id}/latest/cancel`, not in place of workflow ID or run ID which would have violated that assumption. But yes, supporting that and `/api/v1/namespaces/{namespace}/workflows/{workflow_id}/runs/{run_id}/cancel` for every endpoint related to workflows is a bit ugly.

> I'd even be fine with leaving it out entirely and deciding later if we want ?runId=... or a separate route.

I can kinda agree `POST` calls (they can just set it in the body). For getters I think we're leaning to all RPC fields being available as query params, so describing a workflow would probably have this query supported implicitly. But we often offer this as a clear optional value when programmatically interacting with workflows, so it would make sense to be present on the URL.

Reply:
Yikes, my comment above should have said, "If it's [not] something that we expect people to use very often, ..."

Anyway, I'm fine either way; I'll leave this open though in case anyone else wants to chime in.